c++/src/kj/compat/brotli.c++

```diff
@@ -0,0 +1,355 @@
+    }
+
+    if (((peek >> 1) & 0x07) != 0) {
+        return 17 + (peek >> 1 & 0x07);
+    }
+
+    if (((peek >> 4) & 0x07) == 0) {
+        return 17;
+    }
+
+    if (((peek >> 4) & 0x07) == 1) {
+        // Large window brotli, not part of RFC 7932 and not supported in web contexts
+        return BROTLI_MAX_WINDOW_BITS + 1;
+    }
+
+    return 8 + ((peek >> 4) & 0x07);
+}
+
+BrotliOutputContext::BrotliOutputContext(kj::Maybe<int> compressionLevel,
+                                         kj::Maybe<int> _windowBits) : next_in(nullptr), available_in(0) {
+  KJ_IF_MAYBE(level, compressionLevel) {
+    compressing = true;
+    // Emulate zlib's behavior of using -1 to signify the default quality
+    if (*level == -1) {*level = KJ_BROTLI_DEFAULT_QTY;}
+    KJ_REQUIRE(*level >= BROTLI_MIN_QUALITY && *level <= BROTLI_MAX_QUALITY,
+        "invalid brotli compression level", *level);
+    windowBits = _windowBits.orDefault(_::KJ_BROTLI_DEFAULT_WBITS);
+    KJ_REQUIRE(windowBits >= BROTLI_MIN_WINDOW_BITS && windowBits <= BROTLI_MAX_WINDOW_BITS,
+        "invalid brotli window size (window bits", windowBits, ")");
+    cctx = BrotliEncoderCreateInstance(nullptr, nullptr, nullptr);
+    KJ_REQUIRE(cctx, "brotli state allocation failed");
+    // TODO(performance): Make window size configurable from the stream interface and consider
+    // setting up BROTLI_PARAM_LGBLOCK if memory usage becomes an issue
+    KJ_ASSERT(BrotliEncoderSetParameter(cctx, BROTLI_PARAM_QUALITY, *level) == BROTLI_TRUE);
+    KJ_ASSERT(BrotliEncoderSetParameter(cctx, BROTLI_PARAM_LGWIN, windowBits) == BROTLI_TRUE);
+  } else {
+    compressing = false;
+    // In the decoder, we manually check that the stream does not have a higher window size than
+    // requested and reject it otherwise, no way to automate this step.
+    // By default, we accept streams with a window size up to (1 << KJ_BROTLI_MAX_DEC_WBITS),
+    // this is more than the default window size for compression (i.e. KJ_BROTLI_DEFAULT_WBITS)
+    windowBits = _windowBits.orDefault(_::KJ_BROTLI_MAX_DEC_WBITS);
+    KJ_REQUIRE(windowBits >= BROTLI_MIN_WINDOW_BITS && windowBits <= BROTLI_MAX_WINDOW_BITS,
+        "invalid brotli window size (window bits", windowBits, ")");
+    dctx = BrotliDecoderCreateInstance(nullptr, nullptr, nullptr);
+    KJ_REQUIRE(dctx, "brotli state allocation failed");
+  }
+}
+
+BrotliOutputContext::~BrotliOutputContext() noexcept(false) {
+  compressing ? BrotliEncoderDestroyInstance(cctx) : BrotliDecoderDestroyInstance(dctx);
+}
+
+void BrotliOutputContext::setInput(const void* in, size_t size) {
+  next_in = reinterpret_cast<const byte*>(in);
+  available_in = size;
+}
+
+kj::Tuple<bool, kj::ArrayPtr<const byte>> BrotliOutputContext::pumpOnce(BrotliEncoderOperation flush) {
+  byte* next_out = buffer;
+  size_t available_out = sizeof(buffer);
+  // Brotli does not accept a null input pointer; make sure there is a valid pointer even if we are
+  // not actually reading from it.
```

Comment:
When would this occur?

Reply:
I mostly added this for defensive purposes, but it should be possible if pump() is called before `setInput()`.